### ISSUE TITLE
<simpleError in checkwz(wz, M = M, trace = trace, wzepsilon = control$wzepsilon): NAs found in the working weights variable 'wz'>

### ISSUE THREAD

**Issue by brianpenghe** [cole-trapnell-lab/monocle-release#318]
Hi Cole and Xiaojie,
I was running plot_pseudotime_heatmap. And I got that error message although it indeed gave me a figure. The error message is:
<simpleError in checkwz(wz, M = M, trace = trace, wzepsilon = control$wzepsilon): NAs found in the working weights variable 'wz'>

It looks like [some of your demonstration examples](http://cole-trapnell-lab.github.io/monocle-release/tutorial_data/Olsson_dataset_analysis_final.html) also have that.

And I also noticed that not running estimateDispersion() can generate the same error message without giving me a heatmap.

Any ideas?

I've been running Monocle2

**Comment by MichaelPeibo** [comment]
same error to me.

**Comment by MichaelPeibo** [comment]
Hi @brianpenghe 
I have been using integrated data from seurat, which I guess is the cause of the error.
I replaced `exprs(cds)` with raw counts, then I got through the function.

However, I don't know if raw counts are ok to plot the heatmap.

**Comment by dwucsf** [comment]
@MichaelPeibo, do you have both integrated and original RNA counts within your single CDS?

**Comment by MichaelPeibo** [comment]
@dwucsf No, I don't find slot for storing log normalized data or corrected data, so I only replace it.

**Comment by dwucsf** [comment]
Do you use the corrected, integrated data to plot your pseudotime trajectory and then replace the slot with original RNA count data to do downstream analyses?

**Comment by MichaelPeibo** [comment]
Yes,I  use the corrected, integrated data to plot your pseudotime trajectory(different normalization method produce different trajectory, monocle2's lognorm and vst, so I use Seurat's corrected data).

**Comment by dwucsf** [comment]
@MichaelPeibo, can I see your code for how you are constructing your CDS with integrated data, running monocle to get your pseudotime trajectory, and then replacing the integrated counts with original RNA counts within your CDS?

**Comment by MichaelPeibo** [comment]
```
expr_matrix=object@assays$integrated@scale.data # you can use scale.data or data or counts slot, use counts for original RNA counts
expr_matrix=as.matrix(expr_matrix)
### due to different order of meta info, we need subset dataset from ori, then filter
meta_info=object@meta.data 
gene_annotations=as.data.frame(rownames(object@assays$integrated@scale.data))
colnames(gene_annotations)='gene_short_name'
rownames(gene_annotations)=rownames(object@assays$integrated@scale.data)
pd <- new("AnnotatedDataFrame", data = meta_info)
fd <- new("AnnotatedDataFrame", data = gene_annotations)
identical(colnames(expr_matrix), rownames(pd))
cds<- newCellDataSet(as.matrix(expr_matrix),
    phenoData = pd, featureData = fd)

```
```
cds <- estimateSizeFactors(cds)
cds <- estimateDispersions(cds)
cds <- setOrderingFilter(cds, gene_set) #gene_set can be used from you seurat FindMarker()
cds <- reduceDimension(cds, max_components = 2, method = 'DDRTree', norm_method = 'none',pseudo_expr=0) # this step not using monocle's normalization method
cds <- orderCells(cds)

```
NOTE: I am really the fresh to this pseudotime analysis and I am not sure how to refine some parameters to get a nice trajectory, please take this code critically and share me with your idea if you would like.

**Comment by Sophia409** [comment]
@MichaelPeibo Hi, thank you for sharing your code. I also used  similar code to import the integrated data to monocle to do pseudotime analysis.And I also found that the integrated normalized data can cause the error.But how do you replace exprs(cds) with raw counts and which function do you use?I got raw counts from seurat object and tried to resign it to exprs(monocle cds),but I failed.Below is my error.
```
exprs(cds)[,] <- S4@assays$RNA@data[S4@assays$integrated@var.features,][,]
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘exprs<-’ for signature ‘"CellDataSet", "dgCMatrix"’

```
Could you show me your code?Thanks a lot.

**Comment by MichaelPeibo** [comment]
> ```
> expr_matrix=object@assays$integrated@scale.data # you can use scale.data or data or counts slot, use counts for original RNA counts
> expr_matrix=as.matrix(expr_matrix)
> ### due to different order of meta info, we need subset dataset from ori, then filter
> meta_info=object@meta.data 
> gene_annotations=as.data.frame(rownames(object@assays$integrated@scale.data))
> colnames(gene_annotations)='gene_short_name'
> rownames(gene_annotations)=rownames(object@assays$integrated@scale.data)
> pd <- new("AnnotatedDataFrame", data = meta_info)
> fd <- new("AnnotatedDataFrame", data = gene_annotations)
> identical(colnames(expr_matrix), rownames(pd))
> cds<- newCellDataSet(as.matrix(expr_matrix),
>     phenoData = pd, featureData = fd)
> ```
> 
> ```
> cds <- estimateSizeFactors(cds)
> cds <- estimateDispersions(cds)
> cds <- setOrderingFilter(cds, gene_set) #gene_set can be used from you seurat FindMarker()
> cds <- reduceDimension(cds, max_components = 2, method = 'DDRTree', norm_method = 'none',pseudo_expr=0) # this step not using monocle's normalization method
> cds <- orderCells(cds)
> ```
> 
> NOTE: I am really the fresh to this pseudotime analysis and I am not sure how to refine some parameters to get a nice trajectory, please take this code critically and share me with your idea if you would like.

@Sophia409 just the first line of code, 
`expr_matrix=object@assays$integrated@scale.data` for scale.data
`expr_matrix=object@assays$integrated@counts` for raw, or you may try to extract raw counts from `RNA` slot

**Comment by Sophia409** [comment]
@MichaelPeibo Hi, maybe I didn't express it clearly.I can use raw counts as expr_matrix to make a new celldataset.But this can't solve the problem.
I need to import Seurat3 integrated data to monocle to do pseudotime analysis.But the integrated normalized data can only be used to do pseudotime analysis ,not available for any comparison in different conditons.(satijalab/seurat#1674). Besides,as you said,the normalized data cause some errors in some function(such as plot_pseudotime_heatmap),because it onle accept the raw counts.
So I want to replace exprs(cds) with raw counts after ordercells by using normalized data.Then I can plot genes pseudotime dynamics among different branches and states without altering these branches and states information built on the integrated data.
I got raw counts from seurat object and tried to replace exprs(monocle cds),but it failed.You said you replaced exprs(cds) with raw counts,so I mistakenly assume that you replaced the matrix midway instead of importing a newcelldataset.
In fact,two assay can solve this problem,but monocle2 has no this function and they don't plan to add it at present.
I'm glad to hear your idea.

**Comment by MichaelPeibo** [comment]
> @MichaelPeibo Hi, maybe I didn't express it clearly.I can use raw counts as expr_matrix to make a new celldataset.But this can't solve the problem.
> I need to import Seurat3 integrated data to monocle to do pseudotime analysis.But the integrated normalized data can only be used to do pseudotime analysis ,not available for any comparison in different conditons.([satijalab/seurat#1674](https://github.com/satijalab/seurat/issues/1674)). Besides,as you said,the normalized data cause some errors in some function(such as plot_pseudotime_heatmap),because it onle accept the raw counts.
> So I want to replace exprs(cds) with raw counts after ordercells by using normalized data.Then I can plot genes pseudotime dynamics among different branches and states without altering these branches and states information built on the integrated data.
> I got raw counts from seurat object and tried to replace exprs(monocle cds),but it failed.You said you replaced exprs(cds) with raw counts,so I mistakenly assume that you replaced the matrix midway instead of importing a newcelldataset.
> In fact,two assay can solve this problem,but monocle2 has no this function and they don't plan to add it at present.
> I'm glad to hear your idea.

Now I understand what you described, I remembered I tried at first time in the midway of analysis, but somehow it did not work.

**Comment by ChaoXu1997** [comment]
Have you solved this problem? And I meet the same one

**Comment by Wangzy00** [comment]
I solve this problem by decreasing the gene used for doing heatmap from 377(all of the DEG) to Top 50.
Maybe the number of the genes, or some specific gene cause this problem.

**Comment by dimitrisokolowskei** [comment]
I found the same issue while working with spatial datasets. However, I realized that a lot of the genes, for some reason, in the ```differentialGeneTest``` function had the ```num_cells_expressed``` equals to 0. For example:
```sh
diff_test_res <- differentialGeneTest(HSMM_filtered, fullModelFormulaStr="~sm.ns(Pseudotime)")
sort(diff_test_res$num_cells_expressed)


 [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [58] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [115] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [172] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [229] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [286] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [343] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [400] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [457] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [514] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [571] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [628] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [685] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [742] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [799] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [856] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [913] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [970] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [ reached getOption("max.print") -- omitted 16943 entries ]
```
For some reason, the removal of genes expressed in an arbitrary number of cells were not removed in early steps. So, I just did an easy filtering by stablishing a minimum of 10 cells. 
```sh
library(tidyverse)

diff_test_res <- diff_test_res %>%
  filter(num_cells_expressed > 10)
```
Then, proceeded with some pvalue filtering and the plot ran without errors:
```sh
diff <- new[,c("gene_short_name", "pval", "qval","use_for_ordering")]
diff <- diff[diff$pval < 0.05, ]
diff <- diff[order(diff$qval),]
genelist <- row.names(diff)

heatmap <- plot_pseudotime_heatmap(HSMM[genelist], num_clusters=4, show_rownames=T, return_heatmap=T)
``` 

I hope it helps!

**Comment by NGUaqr** [comment]
> I found the same issue while working with spatial datasets. However, I realized that a lot of the genes, for some reason, in the `differentialGeneTest` function had the `num_cells_expressed` equals to 0. For example:
> 
> ```shell
> diff_test_res <- differentialGeneTest(HSMM_filtered, fullModelFormulaStr="~sm.ns(Pseudotime)")
> sort(diff_test_res$num_cells_expressed)
> 
> 
>  [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [58] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [115] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [172] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [229] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [286] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [343] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [400] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [457] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [514] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [571] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [628] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [685] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [742] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [799] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [856] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [913] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [970] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
>  [ reached getOption("max.print") -- omitted 16943 entries ]
> ```
> 
> For some reason, the removal of genes expressed in an arbitrary number of cells were not removed in early steps. So, I just did an easy filtering by stablishing a minimum of 10 cells.
> 
> ```shell
> library(tidyverse)
> 
> diff_test_res <- diff_test_res %>%
>   filter(num_cells_expressed > 10)
> ```
> 
> Then, proceeded with some pvalue filtering and the plot ran without errors:
> 
> ```shell
> diff <- new[,c("gene_short_name", "pval", "qval","use_for_ordering")]
> diff <- diff[diff$pval < 0.05, ]
> diff <- diff[order(diff$qval),]
> genelist <- row.names(diff)
> 
> heatmap <- plot_pseudotime_heatmap(HSMM[genelist], num_clusters=4, show_rownames=T, return_heatmap=T)
> ```
> 
> I hope it helps!

thanks you. I solved the same issue through this way

**Comment by JWDu0122** [comment]
> I solve this problem by decreasing the gene used for doing heatmap from 377(all of the DEG) to Top 50. Maybe the number of the genes, or some specific gene cause this problem.

Thx! I can run when decreasing the genes used for plot heatmap from 1000 to 350, but i don't understand that's why?

**Comment by dd6d** [comment]
Hi, I also found the same problem when I try to plot heatmap for about 10000 genes, while i found that the error only exit when my gene list contain some specific genes. If the list contain these genes, the error occured(so it isn't caused by the number of list).  However, I didn't find out why only these genes could cause the error. So if someone see the error, you could split your list  and delete these genes,write it down for future check.If someone find out why these genes cause the error,please tell me,thanks.

**Comment by JWDu0122** [comment]
> Hi, I also found the same problem when I try to plot heatmap for about 10000 genes, while i found that the error only exit when my gene list contain some specific genes. If the list contain these genes, the error occured(so it isn't caused by the number of list). However, I didn't find out why only these genes could cause the error. So if someone see the error, you could split your list and delete these genes,write it down for future check.If someone find out why these genes cause the error,please tell me,thanks.

You're right, I've also found that when my gene list contains some specific genes, the error occurs, independent of the ranking. Let me know if you figure out a solution to the problem as well.

**Comment by dd6d** [comment]
> > Hi, I also found the same problem when I try to plot heatmap for about 10000 genes, while i found that the error only exit when my gene list contain some specific genes. If the list contain these genes, the error occured(so it isn't caused by the number of list). However, I didn't find out why only these genes could cause the error. So if someone see the error, you could split your list and delete these genes,write it down for future check.If someone find out why these genes cause the error,please tell me,thanks.
> 
> You're right, I've also found that when my gene list contains some specific genes, the error occurs, independent of the ranking. Let me know if you figure out a solution to the problem as well.

For now, I just delete these genes, hope there are better solution

**Comment by yzz32** [comment]
if I change the parameter
 `trend_formula = '~sm.ns(Pseudotime, df=3) * Branch'` 
to
 `trend_formula = '~sm.ns(Pseudotime, df=3) '`
, this error does not occur;      but the result is different;

**Comment by XieYuchong** [comment]
function (val) 
{
  count <- 0
  firstmsg <- NULL
  for (v in val) {
    if (inherits(v, "try-error")) {
      count <- count + 1
      if (count == 1) 
        firstmsg <- v
    }
  }
  if (count == 1) 
    stop("one node produced an error: ", firstmsg, domain = NA)
  else if (count > 1) 
    stop(count, " nodes produced errors; first error: ", 
      firstmsg, domain = NA)
  val
}

plot_pseudotime_heatmap(cds[row.names(c),], num_clusters = 4,use_gene_short_name = TRUE,show_rownames = TRUE,cores=5)
错误于checkForRemoteErrors(val): 
  one node produced an error: the condition has length > 1
